src/daft-io/src/object_io.rs

```diff
@@ -112,10 +117,203 @@ pub(crate) trait ObjectSource: Sync + Send {
+
+    /// Dispatches a task to visit the specified `path` (a concrete path on the filesystem to either a File or Directory).
+    /// Based on the current glob_fragment being processed (accessible via `glob_fragments[i]`) this task will:
+    ///   1. Perform work to retrieve Files/Directories at (`path` + `glob_fragments[i]`)
+    ///   2. Return results to the provided `result_tx` channel based on the provided glob, if appropriate
+    ///   3. Dispatch additional tasks via `.visit()` to continue visiting them, if appropriate
+    fn visit(
+        result_tx: Sender<super::Result<FileMetadata>>,
+        source: Arc<dyn ObjectSource>,
+        path: &str,
+        glob_fragments: (Arc<Vec<GlobFragment>>, usize),
+        full_glob_matcher: Arc<GlobMatcher>,
+    ) {
+        let path = path.to_string();
+        tokio::spawn(async move {
+            log::debug!(target: "glob", "Visiting '{path}' with glob_fragments: {glob_fragments:?}");
+            let (glob_fragments, i) = glob_fragments;
+            let current_fragment = &glob_fragments[i];
+
+            // BASE CASE: current_fragment is a **
+            // We perform a recursive ls and filter on the results for only FileType::File results that match the full glob
+            if current_fragment.escaped_str() == "**" {
+                let next_level_file_metadata =
+                    source.iter_dir(path.as_str(), Some("/"), None).await;
+
+                match next_level_file_metadata {
+                    Ok(mut next_level_file_metadata) => {
+                        while let Some(fm) = next_level_file_metadata.next().await {
+                            match fm {
+                                Ok(fm) => {
+                                    // Recursively visit each sub-directory, do not increment `i` so as to keep visiting the "**" fragmemt
+                                    if matches!(fm.filetype, FileType::Directory) {
+                                        visit(
+                                            result_tx.clone(),
+                                            source.clone(),
+                                            &fm.filepath,
+                                            (glob_fragments.clone(), i),
+                                            full_glob_matcher.clone(),
+                                        );
+                                    }
+                                    // Return any Files that match
+                                    if full_glob_matcher.is_match(fm.filepath.as_str())
+                                        && matches!(fm.filetype, FileType::File)
+                                    {
+                                        result_tx.send(Ok(fm)).await.expect("Internal multithreading channel is broken: results may be incorrect");
+                                    }
+                                }
+                                Err(super::Error::NotFound { .. }) => {}
```

Comment:
shouldn't we propagate this too?

Reply:
Will address in follow-up PR where we optionally propagate `NotFound` if not in a "wildcard state"